File: python/kalman_2d_point.py
from similari import Point2DKalmanFilter, Point2D

if __name__ == '__main__':
    f = Point2DKalmanFilter()
    state = f.initiate(Point2D(1.0, 2.0))

    for i in range(1, 21):
        state = f.predict(state)
        print("Predicted", state.x(), state.y())

        pt = Point2D(1.0 + i * 0.1, 2.0 + i * 0.1)
        print("Observation:", pt)
        state = f.update(state, pt)

File: benches/kalman_2d_point.rs
#![feature(test)]

extern crate test;

use nalgebra::Point2;
use similari::examples::FeatGen2;
use similari::utils::kalman::kalman_2d_point::Point2DKalmanFilter;
use test::Bencher;

#[bench]
fn kalman_2d_point_100k(b: &mut Bencher) {
    const N: usize = 100_000;
    let f = Point2DKalmanFilter::default();
    let mut pt = FeatGen2::new(-10.0, 2.0, 0.2);

    b.iter(|| {
        let v = pt.next().unwrap().feature().as_ref().unwrap().clone();
        let n = v[0].as_array_ref();

        let mut state = f.initiate(&Point2::from([n[0], n[1]]));
        for _i in 0..N {
            let v = pt.next().unwrap().feature().as_ref().unwrap().clone();
            let n = v[0].as_array_ref();
            state = f.predict(&state);

            let p = Point2::from([n[0], n[1]]);
            state = f.update(&state, &p);
        }
    });
}
